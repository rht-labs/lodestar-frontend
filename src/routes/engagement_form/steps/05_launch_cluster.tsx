import React, { useState } from "react";
import { Alert, Button } from "@patternfly/react-core";
import axios, { AxiosError } from "axios";
import slugProperties from "../../../utilities/slug_properties";
import { AppSettings } from "../../../settings/config";

const LaunchCluster = ({ values }: any) => {
  const [isLoading, setIsLoading] = useState(false);
  const [hasError, setErrorResponse] = useState<AxiosError | null>(null);
  const [success, setSuccessResponse] = useState<boolean | null>(null);
  return (
    <div className="pf-l-bullseye">
      <div className="pf-c-empty-state">
        {success ? (
          <span
            aria-label="thumbs up emoji"
            role="img"
            style={{ fontSize: "5rem" }}
          >
            üëçüèΩ
          </span>
        ) : (
            <span
              aria-label="rocket emoji"
              role="img"
              style={{ fontSize: "5rem" }}
            >
              üöÄ
          </span>
          )}

        <div className="pf-c-empty-state__body">
          <h3>
            {!isLoading ? (
              <span>
                We are about to launch your project cluster with this
                information! Are you ready?
              </span>
            ) : (
                <span>
                  This might take a minute. You may want to get a
                <span role="img" aria-label="coffee emoji">
                    ‚òï
                </span>
                </span>
              )}
          </h3>
        </div>

        {!success ? (
          <Button
            onClick={() => {
              setErrorResponse(null);
              setIsLoading(true);
              axios({
                headers: {
                  "X-APPLICATION-NONSENSE": "sure-you-can-access-stuff-#yolo"
                },
                method: "post",
                url: `${AppSettings.backendUrl}/engagements/create`,
                data: slugProperties(values, [
                  "ocp_sub_domain",
                  "customer_name",
                  "project_name"
                ])
              })
                .then(() => {
                  setSuccessResponse(true);
                })
                .catch(error => {
                  setErrorResponse(error);
                })
                .finally(() => {
                  setIsLoading(false);
                });
            }}
            isDisabled={isLoading}
          >
            {!isLoading ? "Let's Do It!" : "Launching..."}
          </Button>
        ) : (
            <div className="pf-c-empty-state">
              <Alert isInline title="You did it!" variant="success">
                <div>
                  Your cluster should be ready soon at:
                <a
                    href={`${values.ocp_sub_domain}.rht-labs.com`}
                  >{`${values.ocp_sub_domain}.rht-labs.com`}</a>
                </div>
              </Alert>
            </div>
          )}

        {hasError ? (
          <div className="pf-c-empty-state">
            <Alert isInline title="We encountered an error." variant="danger">
              {hasError.message}
            </Alert>
          </div>
        ) : null}
      </div>
    </div>
  );
};
export default LaunchCluster;
