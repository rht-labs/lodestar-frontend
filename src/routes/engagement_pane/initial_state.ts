export const initialState = {
  customer_name: null,
  project_name: null,
  description: null,
  location: null,
  start_date: null,
  end_date: null,
  archive_date: null,
  engagement_users: [],
  engagement_lead_name: null,
  engagement_lead_email: null,
  technical_lead_name: null,
  technical_lead_email: null,
  customer_contact_name: null,
  customer_contact_email: null,
  ocp_cloud_provider_name: null,
  ocp_cloud_provider_region: null,
  ocp_version: null,
  ocp_sub_domain: null,
  ocp_persistent_storage_size: null,
  ocp_cluster_size: null,
};
