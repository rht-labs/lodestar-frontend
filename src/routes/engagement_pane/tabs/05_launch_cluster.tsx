import React, { useState, useContext } from 'react';
import { Alert, Button } from '@patternfly/react-core';
import { AxiosError } from 'axios';
import { slugProperties } from '../../../utilities/slug_properties';
import { EngagementContext } from '../../../context/engagement_context';

export const LaunchCluster = ({ values }: any) => {
  const engagementContext = useContext(EngagementContext);
  const [isLoading, setIsLoading] = useState(false);
  const [hasError, setErrorResponse] = useState<AxiosError | null>(null);
  const [success, setSuccessResponse] = useState<boolean | null>(null);
  return (
    <div className="pf-l-bullseye">
      <div className="pf-c-empty-state">
        {success ? (
          <span
            aria-label="thumbs up emoji"
            role="img"
            style={{ fontSize: '5rem' }}
          >
            üëçüèΩ
          </span>
        ) : (
          <span
            aria-label="rocket emoji"
            role="img"
            style={{ fontSize: '5rem' }}
          >
            üöÄ
          </span>
        )}

        <div className="pf-c-empty-state__body">
          <h3>
            {!isLoading ? (
              <span>
                We are about to launch your project cluster with this
                information! Are you ready?
              </span>
            ) : (
              <span>
                This might take a minute. You may want to get a
                <span role="img" aria-label="coffee emoji">
                  ‚òï
                </span>
              </span>
            )}
          </h3>
        </div>

        {!success ? (
          <Button
            onClick={() => {
              setErrorResponse(null);
              setIsLoading(true);

              engagementContext
                .createEngagementCluster(
                  slugProperties(values, [
                    'ocp_sub_domain',
                    'customer_name',
                    'project_name',
                  ])
                )
                .then(() => {
                  setSuccessResponse(true);
                })
                .catch(error => {
                  setErrorResponse(error);
                })
                .finally(() => {
                  setIsLoading(false);
                });
            }}
            isDisabled={isLoading}
          >
            {!isLoading ? "Let's Do It!" : 'Launching...'}
          </Button>
        ) : (
          <div className="pf-c-empty-state">
            <Alert isInline title="You did it!" variant="success">
              <div>
                Your cluster should be ready soon at:
                <a
                  href={`${values.ocp_sub_domain}.rht-labs.com`}
                >{`${values.ocp_sub_domain}.rht-labs.com`}</a>
              </div>
            </Alert>
          </div>
        )}

        {hasError ? (
          <div className="pf-c-empty-state">
            <Alert isInline title="We encountered an error." variant="danger">
              {hasError.message}
            </Alert>
          </div>
        ) : null}
      </div>
    </div>
  );
};
