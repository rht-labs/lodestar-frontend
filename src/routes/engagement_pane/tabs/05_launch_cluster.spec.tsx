import React from 'react';
import { render } from 'enzyme';
import { LaunchCluster } from './05_launch_cluster';
import { initialState } from '../initial_state';

test('Basic component renders correctly', () => {
  const component = render(<LaunchCluster values={initialState} />);
  expect(component).toMatchSnapshot();
});
