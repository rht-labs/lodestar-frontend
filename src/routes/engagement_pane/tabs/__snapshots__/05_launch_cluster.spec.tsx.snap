// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic component renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "pf-l-bullseye",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "pf-c-empty-state",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "aria-label": "rocket emoji",
              "role": "img",
              "style": "font-size:5rem",
            },
            "children": Array [
              Object {
                "data": "ðŸš€",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "pf-c-empty-state__body",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "attribs": Object {},
                      "children": Array [
                        Object {
                          "data": "We are about to launch your project cluster with this information! Are you ready?",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "span",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                  ],
                  "name": "h3",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "pf-c-button pf-m-primary",
                  "data-ouia-component-type": "PF4/Button",
                  "data-ouia-safe": "true",
                  "type": "button",
                },
                "children": Array [
                  Object {
                    "data": "Let's Do It!",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "data-ouia-component-type": undefined,
                  "data-ouia-safe": undefined,
                  "type": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "data-ouia-component-type": undefined,
                  "data-ouia-safe": undefined,
                  "type": undefined,
                },
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "aria-label": undefined,
              "role": undefined,
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "aria-label": undefined,
              "role": undefined,
              "style": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "pf-c-empty-state__body",
            },
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "attribs": Object {},
                    "children": Array [
                      Object {
                        "data": "We are about to launch your project cluster with this information! Are you ready?",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "span",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                ],
                "name": "h3",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "pf-c-button pf-m-primary",
                "data-ouia-component-type": "PF4/Button",
                "data-ouia-safe": "true",
                "type": "button",
              },
              "children": Array [
                Object {
                  "data": "Let's Do It!",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "data-ouia-component-type": undefined,
                "data-ouia-safe": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "data-ouia-component-type": undefined,
                "data-ouia-safe": undefined,
                "type": undefined,
              },
            },
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "aria-label": "rocket emoji",
                "role": "img",
                "style": "font-size:5rem",
              },
              "children": Array [
                Object {
                  "data": "ðŸš€",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "aria-label": undefined,
                "role": undefined,
                "style": undefined,
              },
              "x-attribsPrefix": Object {
                "aria-label": undefined,
                "role": undefined,
                "style": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "pf-c-button pf-m-primary",
              "data-ouia-component-type": "PF4/Button",
              "data-ouia-safe": "true",
              "type": "button",
            },
            "children": Array [
              Object {
                "data": "Let's Do It!",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "pf-c-empty-state__body",
              },
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "attribs": Object {},
                      "children": Array [
                        Object {
                          "data": "We are about to launch your project cluster with this information! Are you ready?",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "span",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                  ],
                  "name": "h3",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "aria-label": "rocket emoji",
                  "role": "img",
                  "style": "font-size:5rem",
                },
                "children": Array [
                  Object {
                    "data": "ðŸš€",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "aria-label": undefined,
                  "role": undefined,
                  "style": undefined,
                },
                "x-attribsPrefix": Object {
                  "aria-label": undefined,
                  "role": undefined,
                  "style": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "data-ouia-component-type": undefined,
              "data-ouia-safe": undefined,
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "data-ouia-component-type": undefined,
              "data-ouia-safe": undefined,
              "type": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
